### PR TITLE
Fix/security issues

### PR DESCRIPTION
**Pull request by cupicmarko**
# Summary

Security hardening and structured error handling across the Rust Soroban UltraHonk verifier and its downstream smart contracts.

The verifier crate replaces panics and opaque `Option` returns with typed `Result` types, adds defense-in-depth validation at every deserialization boundary, and documents cryptographic constant provenance. The three Soroban contracts (identity, generic ultrahonk, tornado) gain reinitialization guards, deploy-time VK validation, governor authorization (identity only), and extensive negative test coverage.

# Changes

- **Fix: Structured VK loading errors**
  - Introduce `VkLoadError` enum (`WrongLength`, `InvalidParameters`) in `verifier.rs`.
  - `utils.rs::load_vk_from_bytes` returns `Result<VerificationKey, VkLoadError>` instead of `Option<VerificationKey>`.
  - `verifier.rs::UltraHonkVerifier::new` returns `Result<Self, VkLoadError>` instead of `Result<Self, VerifyError>`.
  - `utils.rs::load_proof` returns `Result<Proof, &'static str>` instead of panicking on malformed input.
  - `transcript.rs::generate_transcript` returns `Result<Transcript, &'static str>` with up-front `validate_proof` size checks.
  - `VerifyError` simplified to unit variants (`InvalidInput`, `SumcheckFailed`, `ShplonkFailed`), dropping string payloads.

- **Fix: Defense-in-depth arithmetic and deserialization validation**
  - `field.rs::batch_inverse` checks for zero accumulated product before calling `inverse()`.
  - `shplemini.rs` validates `log_circuit_size` range and defensively checks for zero inverted results after batch inversion.
  - `sumcheck.rs` validates `log_circuit_size` range and short-circuits `compute_next_target_sum` at domain points to avoid division-by-zero in the barycentric formula.
  - `utils.rs::load_vk_from_bytes` adds structural parameter validation: `log_circuit_size` bounds, `public_inputs_size >= PAIRING_POINTS_SIZE`, `circuit_size == 1 << log_circuit_size`, `pub_inputs_offset <= circuit_size`.

- **Fix: Contract initialization hardening**
  - Identity, UltraHonk, and Tornado contracts add `AlreadyInitialized` error and guard `__constructor` against double-deployment.
  - Identity contract requires a `governor: Address` parameter, enforces authorization via `require_auth`, stores the governor on-chain, and exposes `governor()` getter.
  - Identity and UltraHonk constructors validate VK bytes at deploy time, mapping `VkLoadError` to distinct contract errors (`VkInvalidLength`, `VkInvalidParameters`).
  - Tornado mixer replaces generic `VerificationFailed` with explicit `InvalidPublicInputs` when public input parsing fails.

- **Doc: Provenance and trust model**
  - `ec.rs` adds provenance comments for BN254 G2 constants (`RHS_G2_BYTES`, `LHS_G2_BYTES`) citing Barretenberg v0.82.2 source.
  - Identity contract README adds Trust Model section documenting governor responsibility, immutability, and user verification requirements.
  - UltraHonk contract adds explicit trust model warning: no governor or access controls; callers must independently verify the VK.

- **Refactor: Internal cleanup**
  - Fix typo: `accumulate_auxillary_relation` renamed to `accumulate_auxiliary_relation`.
  - Make deserialization helpers (`read_bytes`, `combine_limbs`, `fr_word32`, etc.) `pub(crate)`.
  - Tornado mixer `set_root` test helper restricted from `#[cfg(any(test, feature = "testutils"))]` to `#[cfg(test)]`.

- **Test: Negative and edge-case coverage**
  - Verifier crate: `VkLoadError` variant tests, public-input alignment/length tests, VK structural edge cases (zero log, oversized log, size mismatch, bad offset), cross-circuit proof/VK confusion test.
  - Identity contract: empty VK, truncated VK, invalid parameters, double initialization, bad proof length, mutated proof.
  - UltraHonk contract: same constructor and verify-method negative tests.
  - Tornado mixer: short and long public inputs rejection tests.

- **Infra: Deployment and invocation tooling**
  - `justfile` and `scripts/run_identity_e2e.sh` resolve the source account as governor and pass `--governor` to the constructor.
  - New `scripts/verify_identity.sh` standalone script for on-chain identity verification with artifact validation and optional cost measurement.

# Breaking changes

- `UltraHonkVerifier::new` return type changed from `Result<Self, VerifyError>` to `Result<Self, VkLoadError>`. Callers must match `VkLoadError::WrongLength` and `VkLoadError::InvalidParameters`.
- `load_vk_from_bytes` return type changed from `Option<VerificationKey>` to `Result<VerificationKey, VkLoadError>`.
- `load_proof` return type changed from `Proof` to `Result<Proof, &'static str>`.
- `generate_transcript` return type changed from `Transcript` to `Result<Transcript, &'static str>`.
- `VerifyError` variants are now unit variants; string payloads removed.
- Contract error codes renumbered. `VkParseError` removed; replaced by `VkInvalidLength` (1) and `VkInvalidParameters` (2). `ProofParseError` → 3, `VerificationFailed` → 4, `VkNotSet` → 5, `AlreadyInitialized` → 6.
- Identity contract `__constructor` signature changed: now requires `governor: Address` as the first parameter.

### CODE REVIEW

**Comment by maksimryndin** [comment]
@codex review in detail for any vulnerabilities and security issues

File: contracts/identity/src/lib.rs [hunk 1]
```diff
@@ -21,11 +41,41 @@ impl IdentityContract {
         symbol_short!("vk")
     }
 
-    pub fn __constructor(env: Env, vk_bytes: Bytes) -> Result<(), Error> {
+    fn key_governor() -> Symbol {
+        symbol_short!("gov")
+    }
+
+    pub fn __constructor(env: Env, governor: Address, vk_bytes: Bytes) -> Result<(), Error> {
+        if env.storage().instance().has(&Self::key_vk()) {
```
Comment (Oghma) [inline contracts/identity/src/lib.rs:42]: This check is not needed. The constructor is executed once when the smart contract is deployed. Hence the storage is empty. We can keep it as a defensive approach

File: contracts/identity/src/lib.rs [hunk 2]
```diff
@@ -21,11 +41,41 @@ impl IdentityContract {
         symbol_short!("vk")
     }
 
-    pub fn __constructor(env: Env, vk_bytes: Bytes) -> Result<(), Error> {
+    fn key_governor() -> Symbol {
+        symbol_short!("gov")
+    }
+
+    pub fn __constructor(env: Env, governor: Address, vk_bytes: Bytes) -> Result<(), Error> {
```
Comment (Oghma) [inline contracts/identity/src/lib.rs]: I don't understand why add a `governor`. The smart contract is deployed with the verifier key and it's immutable

File: crates/ultrahonk-soroban-verifier/src/sumcheck.rs [hunk 3]
```diff
@@ -85,6 +85,15 @@ fn compute_next_target_sum(
     one: &Fr,
     zero: &Fr,
 ) -> Result<Fr, &'static str> {
+    // Short-circuit: if round_challenge equals any domain point, return the
+    // corresponding univariate value directly. This matches BB behavior and
+    // avoids a division-by-zero in the barycentric formula.
+    for i in 0..BATCHED_RELATION_PARTIAL_LENGTH {
+        if round_challenge == point_indices[i] {
```
Comment (maksimryndin) [inline crates/ultrahonk-soroban-verifier/src/sumcheck.rs]: if index is out of range, there will be a panic

File: contracts/rs-soroban-ultrahonk/src/lib.rs [hunk 4]
```diff
@@ -23,11 +42,28 @@ impl UltraHonkVerifierContract {
     }
 
     /// Initialize the on-chain VK once at deploy time.
+    /// Validates the VK bytes by parsing them before storage so that empty or
+    /// malformed VKs are rejected at deployment time.
     pub fn __constructor(env: Env, vk_bytes: Bytes) -> Result<(), Error> {
+        if env.storage().instance().has(&Self::key_vk()) {
```
Comment (Oghma) [inline contracts/rs-soroban-ultrahonk/src/lib.rs:48]: Same as for the identity contract

File: crates/ultrahonk-soroban-verifier/src/shplemini.rs [hunk 5]
```diff
@@ -93,6 +97,11 @@ pub fn verify_shplemini(
         "shplemini: batch inversion failed (zero denominator in shplonk/gemini/fold)"
     })?;
 
+    // Defense-in-depth: ensure no inverted result is zero before use.
+    if inverted[..batch_size].iter().any(|x| x.is_zero()) {
```
Comment (maksimryndin) [inline crates/ultrahonk-soroban-verifier/src/shplemini.rs]: how batch_size is ensured to be less than inverted.len()?

File: crates/ultrahonk-soroban-verifier/src/utils.rs [hunk 6]
```diff
@@ -194,50 +200,63 @@ pub fn load_vk_from_bytes(env: &Env, bytes: &Bytes) -> Option<VerificationKey> {
     let public_inputs_size = read_u64(bytes, &mut idx);
     let pub_inputs_offset = read_u64(bytes, &mut idx);
 
+    // Validate structural parameters immediately after parsing.
+    if log_circuit_size == 0 || log_circuit_size > CONST_PROOF_SIZE_LOG_N as u64 {
```
Comment (maksimryndin) [inline crates/ultrahonk-soroban-verifier/src/utils.rs]: I would avoid `as` conversions but here it shouldn't be a problem. Anyway explicit fallible conversions are preferred

**Review by maksimryndin — APPROVED** [review]
Awesome @cupicmarko ! 👍 
I've made a very shallow review as I am not in the context but looks good!
Some small comments at your judgement

File: crates/ultrahonk-soroban-verifier/src/utils.rs [hunk 7]
```diff
@@ -45,7 +45,8 @@ const _: () = assert!(
 ///
 /// BB: `field_conversion::calc_num_bn254_frs` + native serialization
 #[inline]
-pub fn coord_to_halves_be(coord: &[u8]) -> ([u8; 32], [u8; 32]) {
+#[allow(dead_code)]
```
Comment (Oghma) [inline crates/ultrahonk-soroban-verifier/src/utils.rs]: Can be removed it if it's dead code?

**Review by Oghma — APPROVED** [review]
I have left some notes. If possible, please remove the dead code before merging

File: scripts/verify_identity.sh [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  echo -e "\n${BLUE}Generating detailed performance report...${NC}"
+  SOURCE_SECRET=$(stellar keys secret "$STELLAR_SOURCE_ACCOUNT" | tail -n 1 | tr -d '[:space:]')
+  SUBMIT_ARG=()
+  if [[ "${MEASURE_SUBMIT:-0}" == "1" ]]; then
+    SUBMIT_ARG=(--submit)
+  fi
+  pushd "$ROOT_DIR/scripts/measure_ultrahonk_costs" >/dev/null
+  npm run measure -- \
+    --contract-id "$CONTRACT_ID" \
+    --source-secret "$SOURCE_SECRET" \
```
[the remark this answers is not in the document]
Comment (cupicmarko) [reply scripts/verify_identity.sh]: Won't fix, scripts are for testing.

File: crates/ultrahonk-soroban-verifier/src/sumcheck.rs (hunk 3 above)
Comment (cupicmarko) [reply crates/ultrahonk-soroban-verifier/src/sumcheck.rs]: Fixed.

File: crates/ultrahonk-soroban-verifier/src/utils.rs (hunk 6 above)
Comment (cupicmarko) [reply crates/ultrahonk-soroban-verifier/src/utils.rs]: Using `u64::try_from` instead, it's just a bit more verbose now.

File: contracts/identity/src/lib.rs (hunk 1 above)
Comment (cupicmarko) [reply contracts/identity/src/lib.rs:42]: Reverted all the changes related to the governor.

File: contracts/identity/src/lib.rs (hunk 2 above)
Comment (cupicmarko) [reply contracts/identity/src/lib.rs]: Reverted all the changes related to the governor.

File: crates/ultrahonk-soroban-verifier/src/utils.rs (hunk 7 above)
Comment (cupicmarko) [reply crates/ultrahonk-soroban-verifier/src/utils.rs]: Moved to test module, it's used there
